CenterForOpenScience/osf.io
# Track Unseen Files [#PLAT-639]

## Purpose

If a user has previously seen a file, and another user uploads a new version of that file, we'd like some indication that the new version is new! So, make the filename bold for new versions of previously viewed files.

<img width="569" alt="screen shot 2018-04-17 at 10 47 56 am" src="https://user-images.githubusercontent.com/801594/38877553-e96790fc-422c-11e8-96f3-f7b7f90a8b1e.png">


## Changes
- Migrations and models for Version - User relationships
- When `create_waterbutler_log` is called and the action is in `DOWNLOAD_ACTIONS`, create a new relationship between the version accessed and the user, store it in `seen_by` on the osfstorage file version
- Return `latestVersionSeen` in `osfstorage_get_children` so that waterbutler can then also pass along this information when rendering the file trees
- Update existing tests for wb callbacks 
- Add tests to make sure relationship is created when file is viewed

## QA Notes
The QA on this one will be interesting! No API changes tho.

It'll work best with 2 users, I'll call them Gertrude and Sally.

1. Sally and Gertrude can both see a project -- they can be collaborators or not
2. Sally uploads a file to the project.
3.  Gertrude goes to the project, sees nothing different about anything
4. Gertrude clicks on the file to go to the file detail page -- she is now aware of the file's existence
5. Sally uploads a new version of the file
6. Gertrude goes back to either the project page or the main file overview page
7. Gertrude will notice that the file she had seen before now has its **filename in bold**
8. Gertrude views the file detail page for that new file version
9. Gertrude goes back to the project overview/file overview page, filename is no longer bold

Notes:
- this also works for non-contributors on projects
- the bold filename should show up on project page, file detail page, and file overview page



## Documentation

- No changes to the API,  internally tried to update new/existing methods with docstrings


## Side Effects
Now sending and receiving new metadata to WB! Shouldn't really affect anything tho.

Here's the blocking WB PR:
https://github.com/CenterForOpenScience/waterbutler/pull/337

## Ticket
Platform ticket:
https://openscience.atlassian.net/browse/PLAT-639
blocking WB ticket:
https://openscience.atlassian.net/browse/SVCS-689
